### PR TITLE
Connection to database

### PR DESCRIPTION
**Pull request by njvision**
Screenshot 2024-07-27 203303
Screenshot 2024-07-27 203135

### CODE REVIEW

File: src/main/java/com/api/entity/SxCompositeFairing.java [hunk 1]
```diff
@@ -2,17 +2,34 @@
...
 import lombok.Getter;
 import lombok.NoArgsConstructor;
 
 @AllArgsConstructor
 @NoArgsConstructor
 @Getter
+@Entity
+@Table(name="sx_composite_fairing")
 public class SxCompositeFairing {
+    @Id
```
Comment (drogojinaru) [inline src/main/java/com/api/entity/SxCompositeFairing.java]: never use same objects for fetching and storing, every operation should have thei objects, please create a separate set of JPA object for the db, SxCompositeFairingJpa for example

File: src/main/resources/db/migration/V2__Insert_SxRocket_Data.sql [hunk 2]
```diff
@@ -0,0 +1,117 @@
+-- Insert data into sx_dimension_rocket
+INSERT INTO sx_dimension_rocket (id, meters, feet)
```
Comment (drogojinaru) [inline src/main/resources/db/migration/V2__Insert_SxRocket_Data.sql]: you need to insert the values within the code, after you fetch them from the API

File: src/main/java/com/api/mapper/RocketMapper.java [hunk 3]
```diff
@@ -54,4 +56,6 @@ public interface RocketMapper {
     Payloads toDto(SxPayloads payloads);
 
     CompositeFairing toDto(SxCompositeFairing compositeFairing);
+
+    FlickrImages toDto(SxFlickrImages flickrImages);
```
Comment (drogojinaru) [inline src/main/java/com/api/mapper/RocketMapper.java]: so basically on namings you will need to have 3 entity types at this moment: RocketDto (used as response in Controller), RocketSx (Used when fetching the model from api), RocketJpa (used as table_entity in DB)

**Comment by njvision** [comment]
Lombok didn't work with mapstruct so getters and setters save save me )

File: src/main/java/com/api/dto/Engines.java [hunk 4]
```diff
@@ -4,10 +4,12 @@
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 import lombok.NoArgsConstructor;
+import lombok.Setter;
 
 @AllArgsConstructor
 @NoArgsConstructor
 @Getter
+@Setter
```
Comment (drogojinaru) [inline src/main/java/com/api/dto/Engines.java:12]: why do you need the setters?

File: src/main/java/com/api/dto/Engines.java [hunk 5]
```diff
@@ -4,10 +4,12 @@
...
 public class Engines {
```
Comment (drogojinaru) [inline src/main/java/com/api/dto/Engines.java]: where are this object used btw? I think we should have only DTO, SX and JPA models

File: src/main/java/com/api/mapper/RocketJpaMapper.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import com.api.model.MassRocketJpa;
+import com.api.model.PayloadWeightsJpa;
+import com.api.model.PayloadsJpa;
+import com.api.model.RocketJpa;
+import com.api.model.SecondStageJpa;
+import com.api.model.ThrustStageJpa;
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring")
+public interface RocketJpaMapper {
```
Comment (drogojinaru) [inline src/main/java/com/api/mapper/RocketJpaMapper.java]: you can divide the Mappers into JPARocket Mapper and hold only the transformation INTO JPA models. And btw, when are you using the case for JPA to SX transformation? you should have JPA to DTO

File: src/main/java/com/api/mapper/RocketMapper.java [hunk 7]
```diff
@@ -26,32 +28,64 @@
...
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.springframework.stereotype.Component;
 
-@Mapper
+@Mapper(componentModel = "spring")
 public interface RocketMapper {
     RocketDto toDto(SxRocket sxRocket);
 
+    SxRocket toDto(RocketDto sxRocket);
```
Comment (drogojinaru) [inline src/main/java/com/api/mapper/RocketMapper.java]: same here, why you have SX to DT added? as the flow is Controller -> fetch from Rocket API -> Save to DB -> Get from DB -> DTO response in controller

File: src/main/java/com/api/model/RocketJpa.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+@Entity
+@Table(name = "rocket")
+public class RocketJpa {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    private Integer id;
+
+//    @Column(name = "_id")
```
Comment (drogojinaru) [inline src/main/java/com/api/model/RocketJpa.java]: commented code

File: src/main/java/com/api/service/NetworkingService.java [hunk 9]
```diff
@@ -30,17 +36,22 @@ public NetworkingService(WebClient webClient) {
     }
 
     public List<SxRocket> getRockets(Integer limit, Integer offset) {
+
+        System.out.println("Start getRockets()");
```
Comment (drogojinaru) [inline src/main/java/com/api/service/NetworkingService.java]: usually use log4j to log the operation, where you can dictate the level of logging

File: src/main/java/com/api/service/RocketService.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import com.api.dto.RocketDto;
+import com.api.entity.SxRocket;
+import com.api.mapper.RocketMapper;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class RocketService {
```
Comment (drogojinaru) [inline src/main/java/com/api/service/RocketService.java]: what is the purpose of this service? is it just doing the mapping? you can do that without separation in other Service

File: src/main/java/com/api/controller/RocketController.java [hunk 11]
```diff
@@ -15,9 +19,15 @@
 public class RocketController {
 
     private final RocketPresenter rocketPresenter;
+    private final RocketPersistenceService rocketPersistenceService;
+    private final RocketService rocketService;
```
Comment (drogojinaru) [inline src/main/java/com/api/controller/RocketController.java]: please separate here the services, controller should be aware only of 1 service or presenter, you should have Controler -> Presenter-> Presenter is responsible of doing the logic, so he is aware of Service, and Persistence, Fetchers etc. , same in the presenter we are making the mapping onto the DTO and sending the response back to the Controller

File: src/main/java/com/api/controller/RocketController.java [hunk 12]
```diff
@@ -26,7 +36,28 @@ public List<RocketDto> getAllRockets(
             @RequestParam(value = "limit", required = false) Integer limitParam,
             @RequestParam(value = "offset", required = false) Integer offsetParam
     ) {
-        return rocketPresenter.getAllRockets(idParam, limitParam, offsetParam);
+
+        List<RocketDto> rocketDtoList = rocketPresenter.getAllRockets(idParam, limitParam, offsetParam);
+        List<SxRocket> sxRocketList = rocketService.getRockets(rocketDtoList);
+        rocketPersistenceService.saveRockets(sxRocketList);
```
Comment (drogojinaru) [inline src/main/java/com/api/controller/RocketController.java]: this should be in the controller, Controller is reponsible only for the UI layer and UI updates/checks

File: src/main/java/com/api/controller/RocketController.java [hunk 13]
```diff
@@ -26,7 +36,28 @@ public List<RocketDto> getAllRockets(
...
+        return rocketDtoList;
+    }
+
+    @GetMapping("/filter")
+    public List<RocketDto> getRockets(
+            @RequestParam(value = "page", defaultValue = "0") int page,
+            @RequestParam(value = "limit", defaultValue = "10") int limit,
+            @RequestParam(value = "offset", defaultValue = "0") int offset) {
```
Comment (drogojinaru) [inline src/main/java/com/api/controller/RocketController.java]: why do you need both page and offset? they can be derived, skip to the usage of 1 (offset for example)

File: src/main/java/com/api/controller/RocketController.java [hunk 14]
```diff
@@ -26,7 +36,28 @@ public List<RocketDto> getAllRockets(
             @RequestParam(value = "limit", required = false) Integer limitParam,
             @RequestParam(value = "offset", required = false) Integer offsetParam
     ) {
-        return rocketPresenter.getAllRockets(idParam, limitParam, offsetParam);
+
+        List<RocketDto> rocketDtoList = rocketPresenter.getAllRockets(idParam, limitParam, offsetParam);
```
Comment (drogojinaru) [inline src/main/java/com/api/controller/RocketController.java]: see the comment above for the presenter

File: src/main/java/com/api/dto/Engines.java (hunk 4 above)
Comment (njvision) [reply src/main/java/com/api/dto/Engines.java:12]: when is called endpoint to get entities from the database, JPA is converted to DTO through the mapper without @Setter mapper cannot set values to DTO from JPA

File: src/main/java/com/api/service/RocketPersistenceService.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+    private final JPARocketMapper JPARocketMapper;
+    private final RocketRepository rocketRepository;
+
+    public RocketPersistenceService(JPARocketMapper JPARocketMapper, RocketRepository rocketRepository) {
+        this.JPARocketMapper = JPARocketMapper;
+        this.rocketRepository = rocketRepository;
+    }
+
+    public void saveRockets (List<RocketDto> rockets) {
```
Comment (drogojinaru) [inline src/main/java/com/api/service/RocketPersistenceService.java]: empty space after saveRockets

**Review by drogojinaru — APPROVED** [review]

**Pull request merged**